sched/irq/irq_attach_thread.c

```diff
@@ -0,0 +1,215 @@
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ ****************************************************************************/
+
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+
+#include <nuttx/config.h>
+
+#include <errno.h>
+#include <stdio.h>
+
+#include <nuttx/irq.h>
+#include <nuttx/arch.h>
+#include <nuttx/kthread.h>
+
+#include "irq/irq.h"
+#include "sched/sched.h"
+
+/****************************************************************************
+ * Privte Types
+ ****************************************************************************/
+
+/* This is the type of the list of interrupt handlers, one for each IRQ.
+ * This type provided all of the information necessary to irq_dispatch to
+ * transfer control to interrupt handlers after the occurrence of an
+ * interrupt.
+ */
+
+struct irq_thread_info_s
+{
+  xcpt_t handler;     /* Address of the interrupt handler */
+  FAR void *arg;      /* The argument provided to the interrupt handler. */
+  sem_t *sem;         /* irq sem used to notify irq thread */
+};
+
+/****************************************************************************
+ * Private Data
+ ****************************************************************************/
+
+static int g_irq_thread_vector[NR_IRQS];
+
+/****************************************************************************
+ * Private Functions
+ ****************************************************************************/
+
+/* Default interrupt handler for threaded interrupts.
+ * Useful for oneshot interrupts.
+ */
+
+static int irq_default_handler(int irq, FAR void *regs, FAR void *arg)
+{
+  FAR struct irq_thread_info_s *irq_info = arg;
+  int ret = IRQ_WAKE_THREAD;
+  xcpt_t vector;
+
+  vector = irq_info->handler;
+
+  if (vector)
+    {
+      ret = vector(irq, regs, irq_info->arg);
+    }
+
+  if (ret == IRQ_WAKE_THREAD)
+    {
+      nxsem_post(irq_info->sem);
+      ret = OK;
+    }
+
+  return ret;
+}
+
+static int isr_thread_main(int argc, FAR char *argv[])
+{
+  unsigned int irq = atoi(argv[1]);
+  xcpt_t vector = (xcpt_t)((uintptr_t)strtoul(argv[2], NULL, 16));
+  xcpt_t handler = (xcpt_t)((uintptr_t)strtoul(argv[3], NULL, 16));
+  FAR void *arg = (FAR void *)((uintptr_t)strtoul(argv[4], NULL, 16));
+  struct irq_thread_info_s irq_info;
+  sem_t sem;
+
+  irq_info.sem = &sem;
+  irq_info.arg = arg;
+  irq_info.handler = handler;
+
+  nxsem_init(&sem, 0, 0);
+
+  irq_attach(irq, irq_default_handler, &irq_info);
+
+#if !defined(CONFIG_ARCH_NOINTC)
+  up_enable_irq(irq);
```

Comment:
why enable irq here? irq_attach_thread should keep the same behavior as irq_attach which doesn't enable irq internally.

Reply:
Due to the optimization of g_irq_thread_vector, results in a delayed invocation of irq_attach. Consequently, the up_enable_irq must also be delayed. A note has been added to document this behavior.